rfcs/20210119-determinism.md

```diff
@@ -0,0 +1,126 @@
+
+Stateless random functions, such as `tf.random.stateless_normal`, are always deterministic and so will never raise determinism-related errors.
+
+No error will be raised if a random op or generator is run before determinism is enabled (as is true for any other op), so users should take care to enable determinism before running any random ops or generators.
+
+### Testing plan
+
+We must ensure that every op will either run deterministically or raise an error if `enable_deterministic_ops` has been called. In order to do this, we must test for determinism, both to check we don't accidentally miss any nondeterministic ops and to ensure any previously-deterministic ops are not accidentally made nondeterministic in the future. To accomplish this, we will have the following tests:
+
+1. We will add tests to several of the [official models](https://github.com/tensorflow/models/tree/master/official) to ensure they run deterministically. In particular, each test will train a model for several steps, then retrain it from scratch several times. The final weights after training will be asserted to be the same each time. This tests not only the `enable_deterministic_ops` API but that the entire model is deterministic. This only tests ops that the official models use.
+
+2. We will add a special mode to TensorFlow where every time a non-stateful op is run, TensorFlow will rerun the op several times and assert the outputs are the same each time. We will then run the TensorFlow unit tests with this mode as part of the nightly tests. Doing so ensures that for each op that is run as part of a unit test, it will be tested for determinism.
+
+3. When adding determinism to an op which previously was nondeterministic, an explicit unit test will be added that checks for determinism. Unlike running unit tests with the special mode above, the explicit unit tests can be part of the presubmit tests instead of the nightly tests, and can test on inputs that are very likely to demonstrate nondeterminism if it exists.
+
+### Op Review and changes
+
+As part of the implementation, we will review all ops to make a determination of their behavior (deterministic vs nondeterministic). Ops that are known to operate nondeterministically, at least when running on a GPU, include the following:
+
+* `tf.nn.softmax_cross_entropy_with_logits`
+* `tf.nn.sparse_softmax_cross_entropy_with_logits`
+* `tf.image.resize` gradient with `method=ResizeMethod.NEAREST`
+* `tf.math.segment_sum`, `tf.math.unsorted_segment_sum` forward
+* `tf.image.crop_and_resize` gradient to both image and boxes 
+* `tf.math.unsorted_segment_mean`, `tf.math.unsorted_segment_prod` and `tf.math.unsorted_segment_sqrt`; all forward
+* `tf.sparse.sparse_dense_matmul`
+* `tf.image.adjust_contrast`
+* `tf.compat.v1.nn.fused_batch_norm` gradient w.r.t. `offset`
```

Comment:
I propose replacing the existing list with the following list. This updated list adds ops, fixes various issues, and sorts the ops lexicographically.

```
* `tfa.image.dense_image_warp` gradient w.r.t `image`
* `tf.compat.v1.nn.fused_batch_norm` gradient w.r.t. `offset`
* `tf.convert_to_tensor` forward, when `value` is of type `tf.IndexedSlices`
* `tf.gather` gradient w.r.t dense `params`
* `tf.image.resize` gradient w.r.t `image`, when `method=ResizeMethod.NEAREST`
* `tf.image.crop_and_resize` gradient w.r.t both `image` and `boxes`
* `tf.image.adjust_contrast` forward
* `tf.math_segment_prod` forward
* `tf.math.segment_sum` forward
* `tf.math.unsorted_segment_mean` forward
* `tf.math.unsorted_segment_sqrt_n` forward
* `tf.math.unsorted_segment_prod` forward
* `tf.math.unsorted_segment_sum` forward
* `tf.nn.softmax_cross_entropy_with_logits` forward
* `tf.nn.sparse_softmax_cross_entropy_with_logits` forward
* `tf.sparse.sparse_dense_matmul` forward
```

Reply:
Oops. Sorry, I missed that this was merged. I will create a PR for the above.